cachi2/core/package_managers/yarn_classic/resolver.py

```diff
@@ -27,47 +37,137 @@
     re.compile(r"^https?:.+\.git#.+"),
 )
 
+YARN_REGISTRY_URL = "https://registry.yarnpkg.com"
+
 
-class _BasePackage(BaseModel):
+class _BasePackage(BaseModel, ABC):
     """A base Yarn 1.x package."""
 
     name: str
     version: Optional[str] = None
     integrity: Optional[str] = None
     dev: bool = False
 
+    @property
+    @abstractmethod
+    def purl(self) -> str:
+        """Return the package URL."""
+
 
 class _UrlMixin(BaseModel):
     url: str
 
 
 class _RelpathMixin(BaseModel):
+    root: RootedPath
     relpath: Path
```

Comment:
1. Do you think the mixin's name should be slightly reworded ?
2. And what is the more appropriate type for `root` ? `RootedPath` or just `Path`

Reply:
Maybe just `_PathMixin`?

I'm wondering if we could just have one RootedPath and then just use the root and subpath_from_root properties of that one RootedPath to get everything we need